src/main/java/picard/sam/SamFormatConverter.java

```diff
@@ -72,12 +68,24 @@ public static void main(final String[] argv) {
     }
 
     protected int doWork() {
-        IOUtil.assertFileIsReadable(INPUT);
-        IOUtil.assertFileIsWritable(OUTPUT);
-        final SamReader reader = SamReaderFactory.makeDefault().referenceSequence(REFERENCE_SEQUENCE).open(INPUT);
-        final SAMFileWriter writer = new SAMFileWriterFactory().makeWriter(reader.getFileHeader(), true, OUTPUT, REFERENCE_SEQUENCE);
+        convert(INPUT, OUTPUT, REFERENCE_SEQUENCE, CREATE_INDEX);
+        return 0;
+    }
 
-        if (CREATE_INDEX && writer.getFileHeader().getSortOrder() != SAMFileHeader.SortOrder.coordinate) {
+    /**
+     * Convert a file from one of sam/bam/cram format to another based on the extension of output.
+     *
+     * @param input             input file in one of sam/bam/cram format
+     * @param output            output to write converted file to, the conversion is based on the extension of this filename
+     * @param referenceSequence the reference sequence to use, necessary when reading/writing cram
+     * @param createIndex       whether or not an index should be written alongside the output file
+     */
+    public static void convert(final File input, final File output, final File referenceSequence, final Boolean createIndex) {
```

Comment:
I extracted the guts of this tool into a public method so it could be tested more sanely.